applications/Polkadot-Protocol-Conformance-Tests.md

```diff
@@ -0,0 +1,163 @@
+
+### Host API
+
+The Host API consists of a set of functions that the Host exposes to the Runtime. These functions are used to access external resources for various purposes, including storage access, manipulation, memory allocation, and more. If a method within the API contains a bug, it has the potential to push the Host into an incorrect state transition. Such a scenario could lead to undefined consequences, particularly if a significant number of nodes experience the same issue.
+
+### SCALE
+
+Substrate employs a lightweight and efficient encoding and decoding mechanism to optimise the transmission and reception of data across the network. This protocol, known as the SCALE codec, plays a vital role in serialising and deserialising data. It serves as a critical component for data transfer across the peer network and facilitates communication between the Runtime and the Host. Consequently, the presence of comprehensive tests for SCALE encoding and decoding holds immense significance in ensuring the proper functionality of the Host.
+
+### State Trie
+
+The state trie is another crucial part of the Host. A radix-16 state is the data structure that Substrate uses to store the state of the blockchain. Without thoroughly tested state trie functionalities, the Host may transition to an incorrect state and get slashed if it's a block producer.
+
+### BABE & GRANDPA
+
+[BABE](https://spec.polkadot.network/sect-block-production) & [GRANDPA](https://github.com/paritytech/finality-grandpa) are the bread and butter of the consensus-reaching module for Polkadot. However, finding the right tools and approach to test the block production and finalisation protocols independently of the Host’s environment is a challenge that is yet to be overcome by any team.
+
+### Zombienet
+
+[Zombienet](https://github.com/paritytech/zombienet) aims to be a testing framework for Substrate-based blockchains, providing a CLI tool that allows users to spawn and test ephemeral networks. The assertions used in the tests can include on-chain storage, metrics, logs, and custom JavaScript scripts that interact with the chain
+
+Our team has successfully utilised Zombienet to run PVF conformance tests as part of another W3F grant. This is the reason why we believe that Zombienet has the potential to be the go-to framework for running conformance tests on the Hosts. The research is going to focus on Zombienet, as well as on the new [Zombienet SDK](https://github.com/paritytech/zombienet-sdk).
```

Comment:
Could you please elaborate on how Zombienet was used to integrate with the PVF conformance test-suite? And briefly compare and contrast with the current project?

Reply:
In the PVF conformance test-suite, we leveraged Zombienet's custom WASM [configuration option](https://paritytech.github.io/zombienet/network-definition-spec.html)(genesis_wasm_path) in order to inject malicious PVFs. Additionally, the team used Zombienet's DSL language in order to assert on the behaviour of the network. You can find the tests [here](https://github.com/LimeChain/polkadot-conformance/tree/main/tests). They're split into different folders and files mainly because it's easier to read and understand what's happening. The main difference between Polkadot and Kagome's tests is the startup command. There are also one or two differences in terms of logs emitted by each Host.

Reply:
When it comes to comparing the 'adapter' approach with the Zombienet approach, the biggest difference is in terms of maintainability and being forward-compatible. The adapter approach is dependent on each Host not changing its implementation details. If the function's name is changed or is extracted in another module/repository, the testing suite should change as well; otherwise, it won't compile. If a function's logic is changed, then this may result in tests failing. These are all cases very relevant for Hosts in early stages of their implementation. Whereas, changes to CLI parameters and public APIs (which Zombienet uses under the hood) happen less often, which makes the approach better in terms of maintainability